core/lib/circuit_breaker/src/lib.rs

```diff
@@ -21,50 +19,64 @@ pub enum CircuitBreakerError {
 /// Checks circuit breakers
 #[derive(Debug)]
 pub struct CircuitBreakerChecker {
-    circuit_breakers: Vec<Box<dyn CircuitBreaker>>,
-    sync_interval: Duration,
+    circuit_breakers: Mutex<Vec<Box<dyn CircuitBreaker>>>,
+    config: CircuitBreakerConfig,
 }
 
 #[async_trait::async_trait]
 pub trait CircuitBreaker: std::fmt::Debug + Send + Sync {
+    fn id(&self) -> &'static str;
+
     async fn check(&self) -> Result<(), CircuitBreakerError>;
 }
 
+impl PartialEq for dyn CircuitBreaker {
+    fn eq(&self, other: &Self) -> bool {
+        self.id() == other.id()
+    }
+}
+
 impl CircuitBreakerChecker {
     pub fn new(
-        circuit_breakers: Vec<Box<dyn CircuitBreaker>>,
+        circuit_breakers: Option<Vec<Box<dyn CircuitBreaker>>>,
         config: &CircuitBreakerConfig,
     ) -> Self {
         Self {
-            circuit_breakers,
-            sync_interval: config.sync_interval(),
+            circuit_breakers: Mutex::new(circuit_breakers.unwrap_or_default()),
+            config: config.clone(),
         }
     }
 
+    pub fn replication_lag_limit_sec(&self) -> Option<u32> {
+        self.config.replication_lag_limit_sec
+    }
+
     pub async fn check(&self) -> Result<(), CircuitBreakerError> {
-        for circuit_breaker in &self.circuit_breakers {
+        for circuit_breaker in self.circuit_breakers.lock().await.iter() {
             circuit_breaker.check().await?;
         }
         Ok(())
     }
 
-    pub async fn run(
-        self,
-        circuit_breaker_sender: oneshot::Sender<CircuitBreakerError>,
-        stop_receiver: watch::Receiver<bool>,
-    ) -> anyhow::Result<()> {
+    pub async fn insert_breaker_if_not_exists(&self, circuit_breaker: Box<dyn CircuitBreaker>) {
+        let mut guard = self.circuit_breakers.lock().await;
+        if !guard.contains(&circuit_breaker) {
+            guard.push(circuit_breaker);
+        }
+    }
+
+    pub async fn run(self: Arc<Self>, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
```

Comment:
Once you split the checker into two parts (see my other comment), you'll be able to get rid of `Arc<Self>` here and make it accept just `self` again.